Docs for `JupyterAnalysis` and `Action` schemas (#77)
FAIRmat-NFDI/nomad-analysis: docs/how_to/analysis_with_jupyter_notebooks.md

Reviewer: Do we also want a brief intro explaining the WHY?
Just something short that highlights semi-automated data analysis and sharing of analysis pipelines all via the schema.

## Why use Jupyter Analysis?

Jupyter notebooks are a powerful tool for interactive data analysis, combining
code, visualizations, and documentation in a single document. The
Jupyter Analysis schema leverages this capability to provide a structured
framework for performing semi-automated analysis workflows directly connected
to your research data.

With Jupyter Analysis entries, you can connect entries stored in
NOMAD as inputs, load them into your notebook environment, and perform custom
analysis using Python. The results can then be saved back to NOMAD, ensuring
that your analysis outputs are properly linked to the original data and remain
part of the reproducible research record. This approach also makes it easy to
share your complete analysis workflow—including code, data references, and
results—with collaborators.

## Method 1: Using the ELN Interface

### Create a new Jupyter Analysis Entry

- Navigate to your upload in NOMAD
- Click **Create Entry** and enter the name of your analysis (e.g., "XRD Phase Analysis")

> - Select the **Built-in Schemas** option and select **Jupyter Analysis** from the dropdown

Reviewer: ```suggestion
- Select the **Built-in Schemas** option and select **Jupyter Analysis** from the dropdown menu
```

- Click **Create** to generate the entry

### Fill in Basic Information

| Field | Description |
|-------|-------------|
| **Name** | A descriptive name for your analysis |
| **Description** | Detailed description of the analysis purpose |
| **Lab ID** | Optional laboratory identifier |

### Select Input Data

> Existing NOMAD entries can be added as input to your analysis via referencing.

Reviewer: Minor grammar correction + typesetting to denote that we're using reference with a very technical meaning.

```suggestion
Existing NOMAD entries can be added as input to your analysis as `Reference`s.
```

To do this, go to the `inputs` field and click on the plus button to add a new
sub-section. Inside the sub-section, fill in the `reference` quantity with path to the desired entry and save the analysis entry.

> If you want to add several inputs, you can also use a query-powered approach to search and reference multiple entries at once. To do this:

Reviewer: Would add emphasis for visibility:

```suggestion
If you want to add several inputs, you can also use a **query-powered approach** to search and reference multiple entries at once. To do this:
```

- Click OK to add the searched entries as inputs

Each searched entry will be processed as one input and all of them will be added under the `inputs` field.

!!! hint
    If you have mistakenly added a lot of inputs based on a broad query, use the **Reset Inputs** action button to start afresh. Set the more specific query and click the button. All existing inputs will be removed and repopulated based on the new query.

### Generate the Notebook

A notebook can be uploaded manually and connected to the entry. For this, simply find the `notebook` quantity and upload a Jupyter notebook file.

You can also generate a notebook automatically by clicking the **Generate Notebook** action button. This will:

- Create a new `.ipynb` file in the same upload
- Pre-populate it with header cells and data loading code
- Link it with the `notebook` quantity

The name of the generated notebook will match your entry name. It contains a header with the analysis name and a pre-defined code cell that loads the analysis entry and its inputs:

```python
from nomad_analysis.utils import get_entry_data

analysis = get_entry_data(entry_id=<NOMAD_ANALYSIS_ENTRY_ID>)
```

> Inputs can be accessed through the `analysis.inputs` attribute, which contains references to the input entries you defined in the ELN interface.

Reviewer: I think this phrasing is a bit dangerous, as `analysis.input` refers to the variable name in your example exclusively.

Either use `get_entry_data(entry_id=<NOMAD_ANALYSIS_ENTRY_ID>).inputs`, or denote the class of the object `get_entry_data` returns. You can also use Python type hints for this in your example.

!!! warning
    Different notebooks cannot share the same name. If you encounter issues trying to generate a new notebook, check for name collisions first. Either modify or delete the existing notebook depending on your needs.

### Open and Run the Notebook

You can open the notebook in NOMAD's integrated JupyterHub (North).

- Click on right arrow button next to the `notebook` field
- From a list of NORTH tools, select the JupterHub by clicking on the "Launch" button under it

> It will open in NOMAD's integrated JupyterHub (North) in a new tab with your notebook loaded. The directory structure in the JupyterHub will mirror your NOMAD upload structure, allowing you to easily access your data files and notebooks.

Reviewer: ```suggestion
NOMAD's integrated JupyterHub (North) will automatically open in a new tab with your notebook loaded. The directory structure in the JupyterHub will mirror your NOMAD upload structure, allowing you to easily access your data files and notebooks.
```

Now you can run the cells to load your data and perform your analysis.

During the analysis, you can save the steps and results by writing them into
the `analysis.steps` and `analysis.outputs` fields respectively. To save the
entry back to NOMAD, simply call `analysis.save()` at the end of your notebook.
This will update the entry with your analysis results and ensure that

> everything is properly linked and stored in NOMAD.

Reviewer: This is really clear and neat. Super-technical question: does saving the results trigger a full reprocessing of the analysis entry or its input entries?

Author: The `save` method uses the PUT request with `uploads/{upload_id}/raw/{path}` entrypoint, which triggers processing, I believe. The method also uses `wait_for_processing` flag as True. But remember, it's the `JupyterAnalysis` entry that will be reprocessed, not the referenced entries.

## Method 2: Using YAML Files

Jupyter Analysis entries can also be created using YAML files. The support
comes from NOMAD in general, where you can create entries by uploading YAML
files with the appropriate structure.

Create a file named `my_analysis.archive.yaml` and upload it to your NOMAD upload:

```yaml
data:
  m_def: nomad_analysis.jupyter.schema.JupyterAnalysis
  name: XRD Phase Analysis
  description: |
    Analysis of XRD patterns to identify crystalline phases
    in the synthesized samples.

  # Optional: Pre-define input references
  inputs:
    - reference: '../uploads/UPLOAD_ID/archive/ENTRY_ID_1#/data'
      name: Sample A - XRD Pattern
    - reference: '../uploads/UPLOAD_ID/archive/ENTRY_ID_2#/data'
      name: Sample B - XRD Pattern
```

After uploading, trigger notebook generation and other functionality through

> the ELN interface (see above).

Reviewer: You could even put this in a warning admonition, as it's important to know that **analysis never runs automaticly**.

## Method 1: Using the ELN Interface

### Create a New Entry

- Navigate to your upload in NOMAD
- Click **Create Entry** and enter the name of your analysis (e.g., "XRD Phase Analysis")
- Select the **Built-in Schemas** option and select **Jupyter Analysis** from the dropdown
- Click **Create** to generate the entry

### Fill in Basic Information

| Field | Description |
|-------|-------------|
| **Name** | A descriptive name for your analysis |
| **Description** | Detailed description of the analysis purpose |
| **Lab ID** | Optional laboratory identifier |

### Select Input Data

You can search and connect multiple input entries to your analysis. To do this in the ELN:

- Find the `query_for_inputs` field
- Click the search icon to open the search interface
- Build a query to find your input data (e.g., filter by upload, entry type, etc.)
- Click OK to add the searched entries as inputs

The entries will be stored and appear in the `inputs` sub-section.

A new input can also be added manually by clicking the plus button in the `inputs` sub-section and filling in the reference path to the desired entry.

> 

Reviewer: Like the structure, would make the 2 options a bit clearer, maybe via subsection headers: query-powered; manual

Author: I agree it's confusing. I will change it to start with manually adding an input reference, and then go on to the query-powered approach. I would avoid creating subsection headers as the content for each isn't much.

### Resetting Inputs

> To clear and repopulate inputs from queries, click the **Reset Inputs** action button. All existing inputs will be removed and repopulated based on the search query in `query_for_inputs`.

Reviewer: Maybe works better as an info or hint admonition instead of a section?

### Generate the Notebook

A notebook can be uploaded manually and connected to the entry. For this, simply find the `notebook` quantity and upload a Jupyter notebook file.

You can also generate a notebook automatically by clicking the **Generate Notebook** action button. This will:

- Create a new `.ipynb` file in the same upload
- Pre-populate it with header cells and data loading code
- Link it with the `notebook` quantity

The name of the generated notebook will match your entry name.

!!! note

>     If a notebook already exists with the same name, the generation will be skipped. Delete the existing notebook first if you want to regenerate it.

Reviewer: ```suggestion
!!! warning
    Different notebooks cannot share the same name. If you encounter issues trying to generate a new notebook, check for name collisions first. Either modify or delete the existing notebook depending on your needs.
```

### Open and Run the Notebook

You can open the notebook in NOMAD's integrated JupyterHub (North).

- Click on right arrow button next to the `notebook` field
- From a list of NORTH tools, select the JupterHub by clicking on the "Launch" button under it

It will open in NOMAD's integrated JupyterHub (North) in a new tab with your notebook loaded. The directory structure in the JupyterHub will mirror your NOMAD upload structure, allowing you to easily access your data files and notebooks.

> Now you can run the cells to load your data and perform your analysis.

Reviewer: This misses a part on how to load in data from the archive into the Jupyter Notebook. If anything is prefilled by us, please mention that (perhaps in the previous section0.

## Method 2: Using YAML Files

> `JupyterAnalysis` entries can also be created using YAML files. The support comes from NOMAD in general, where you can create entries by uploading YAML files with the appropriate structure.

Reviewer: ```suggestion
NOMAD's ad hoc YAML schema (<!-- link -->) also support `JupyterAnalysis` entries.
```

Author: I see your point, but I couldn't find a central documentation section on using YAML for creating entries. Only found one for defining schema and wrote this minimal description.

Reviewer: Ow, that's worrying. Still, I think the rephrasing is clearer, even without a link.

Create a file named `my_analysis.archive.yaml` and upload it to your NOMAD upload:

```yaml
data:
  m_def: nomad_analysis.jupyter.schema.JupyterAnalysis
  name: XRD Phase Analysis
  description: |
    Analysis of XRD patterns to identify crystalline phases
    in the synthesized samples.

  # Optional: Pre-define input references
  inputs:
    - reference: '../uploads/UPLOAD_ID/archive/ENTRY_ID_1#/data'
      name: Sample A - XRD Pattern
    - reference: '../uploads/UPLOAD_ID/archive/ENTRY_ID_2#/data'
      name: Sample B - XRD Pattern
```

After uploading, trigger notebook generation and other functionality through

> the ELN interface.

Reviewer: ```suggestion
After uploading, trigger notebook generation and other functionality through
the ELN interface (see above).
```
